dm/docs/RFCS/20211130_enhanced_pre_checker.md

```diff
@@ -0,0 +1,89 @@
+
+### Too much time overhead
+
+If we have a large number of tables in source, we will take too much time in checking table schema, sharding table consistency and sharding table auto increment ID.
+
+### Inadequate check
+
+* Now we check auto increment ID by `column-mapping` which is deprecated. If we don't set the `column-mapping` and don't ignore the **auto_increment_ID**, the pre-check will report an error. PS: **auto_increment_ID** is only checked when **schema_of_shard_tables** is true.
+* Dump privilege only checks RELOAD and SELECT. However, Dumpling supports different [consistency configurations](https://docs.pingcap.com/tidb/stable/dumpling-overview#adjust-dumplings-data-consistency-options), which need more privileges.
+* If online-ddl is set by true and a DDL is in online-ddl stage, DM will have a problem in all mode. Specifically, when ghost table has been created but not been renamed, DM will report an error when the ghost table is renamed during the incremental phase. You can learn more about online-ddl [here](https://docs.pingcap.com/tidb-data-migration/stable/feature-online-ddl).
+* For schema_of_shard_tables, whether it's a pessimistic task or an optimistic task, we all simply check consistency of schema by comparing all sharding tables’ structures. For optimistic mode, we can do better.
+* Version checker checks that MySQLVersion >= 5.6.0 and MariadbVersion >= 10.1.2 before. However, we find more and more incompatibility problems in MySQLVersion >= 8.0.0 and Mariadb. In view of supporting MySQL 8.0 and Mariadb is experimental yet, checker will report a warning for them.
+
+## Proposal
+
+### Speed ​​up check
+
+1. Support concurrent check
+    - table_schema
+    - schema_of_shard_tables
+    - auto_increment_ID
+2. We casn adjust the concurrency by table numbers.
+
+### Optimize some check
+
+1. **auto_increment_ID** only checked in sharding mode.
+    - If table exist auto increment ID, report a warning to user and tell them the method that can resolve the PK/UK conflict;
+        1. If you set PK to AUTO_INCREMENT, you must make sure that the primary key in sharding tables is not duplicated;
+        2. If sharding tables have duplicated PK, you can refer to [document](https://docs.pingcap.com/tidb-data-migration/stable/shard-merge-best-practices#handle-conflicts-of-auto-increment-primary-key).
+2. Dump_privilege will check different privileges according to different [consistency](https://docs.pingcap.com/tidb/stable/dumpling-overview#adjust-dumplings-data-consistency-options) and downstream on source.
+    - For all consistency, we will check
+        - REPLICATION CLIENT (global)
+        - SELECT (only INFORMATION_SCHEMA's tables and dump tables)
+    - For flush consistency:
+        - RELOAD (global)
+    - For flush/lock consistency:
+        - LOCK TABLES (only dump tables)
+    
+    As we know, TiDB is different from MySQL in some place. If source is TiDB, we also need:
```

Comment:
we have not plan to support TiDB source in my memory 🤔

Reply:
Indeed. 😂 Do we need to do something for TiDB source? @lichunzhu

Reply:
Maybe for all source expect MySQL and MariaDB, we all report a warning? @lance6716